kerastuner/engine/oracle.py

```diff
@@ -17,90 +17,241 @@
 from __future__ import division
 from __future__ import print_function
 
+import collections
 import enum
 import hashlib
+import json
 
+from tensorflow import keras
+from kerastuner.engine import hyperparameters as hp_module
+from kerastuner.engine import metrics_tracking
+from kerastuner.engine import stateful
+from kerastuner.engine import trial as trial_lib
+from ..abstractions.tensorflow import TENSORFLOW_UTILS as tf_utils
 
-class Oracle(object):
 
-    def __init__(self):
-        self.space = []
+Objective = collections.namedtuple('Objective', 'name direction')
 
-    def update_space(self, new_entries):
+
+class Oracle(stateful.Stateful):
+
+    def __init__(self,
+                 objective,
```

Comment:
How does this support multi-objective cases?

Reply:
The base class only supports formatting multi-objective Objectives, not scoring them (too many assumptions need to be made about how to score a multi-objective trial), but single objective is now only assumed in `score_trial` and `get_best_trials` and so subclasses of Oracle can easily support